program-runtime/src/invoke_context.rs

```diff
@@ -769,6 +773,17 @@ impl<'a> InvokeContext<'a> {
                 let post_remaining_units = self.get_remaining();
                 *compute_units_consumed = pre_remaining_units.saturating_sub(post_remaining_units);
 
+                // startign from feature `native_programs_consume_cu`, all builtin programs should consume
+                // compute unitsi, otherwise will result to InstructionError.
+                if is_builtin_program
+                    && *compute_units_consumed == 0
+                    && self
+                        .feature_set
+                        .is_active(&native_programs_consume_cu::id())
+                {
+                    return Err(InstructionError::BuiltinProgramsMustConsumeUnits);
```

Comment:
Now that I think about it more -- consuming 0 units is a *programmer* error and not a user error, so we could be even more aggressive and assert.

On the flipside, since the runtime is used by other developers, like in all of the program tests, they should probably get a nice error rather than an assert. I'm leaning toward keeping this as an error, but I wanted to bring up the discussion to see how you thought.

Reply:
Great observation! I agree it'd be a _programmer_ error. Also agree runtime is library other developers use; so it'd be great, as good SDK, to return meaningful errors for unsupported user invokes. I'd vote for keep error too.